lib/bolt/module_installer/puppetfile.rb

```diff
@@ -0,0 +1,146 @@
+# frozen_string_literal: true
+
+require 'puppetfile-resolver'
+
+require_relative '../error'
+require_relative 'puppetfile/forge_module'
+require_relative 'puppetfile/git_module'
+
+# This class manages the logical contents of a Puppetfile. It includes methods
+# for parsing and generating a Puppetfile.
+#
+module Bolt
+  class ModuleInstaller
+    class Puppetfile
+      attr_reader :modules
+
+      def initialize(modules = [])
+        @modules = []
+        add_modules(modules)
+      end
+
+      # Loads a Puppetfile and parses its modules.
+      #
+      def self.parse(path, skip_unsupported_modules: false)
+        return new unless path.exist?
+
+        begin
+          parsed = ::PuppetfileResolver::Puppetfile::Parser::R10KEval.parse(File.read(path))
+        rescue StandardError => e
+          raise Bolt::Error.new(
+            "Unable to parse Puppetfile #{path}: #{e.message}",
+            'bolt/puppetfile-parsing'
+          )
+        end
+
+        unless parsed.valid?
+          raise Bolt::ValidationError, <<~MSG
+            Unable to parse Puppetfile #{path}:
+            #{parsed.validation_errors.join("\n\n")}.
+            This may not be a Puppetfile managed by Bolt.
+          MSG
+        end
+
+        new(serialize_parsed_modules(parsed.modules, skip_unsupported_modules))
+      end
+
+      # Serializes the modules returned from the parser.
+      #
+      private_class_method def self.serialize_parsed_modules(modules, skip_unsupported)
```

Comment:
Serialization isn't quite what this is doing.

Reply:
`hashify_parsed_modules`? `parsed_modules_to_data`? 🤔

Reply:
I'm not entirely clear on why we turn these into data instead of directly into module objects. It also seems like we only need to turn them into our own module objects because from the puppetfile-resolver they don't have all the attributes we need, but it might be easier to just PR the resolver than do all this munging?

Reply:
The thought was to have a single place where those objects are instantiated, but it doesn't really matter. 🤷 And they're converted to our own module objects because the resolver returns a generic `ModuleSpecification` object that includes a limited set of attributes and does not have methods that are helpful for our purposes such as `to_spec` or `to_hash`.

Reply:
Updated to just create the module objects directly.